crates/metaslang/bindings/src/builder/mod.rs

```diff
@@ -825,20 +793,96 @@ impl<'a, KT: KindTypes> Builder<'a, KT> {
         Ok(())
     }
 
-    fn load_definiens_info(
-        &mut self,
-        node_ref: GraphNodeRef,
-        node_handle: Handle<Node>,
-    ) -> Result<(), BuildError> {
+    fn node_handle_for_graph_node(&self, node_ref: GraphNodeRef) -> Handle<Node> {
+        self.stack_graph
+            .node_for_id(self.node_id_for_graph_node(node_ref))
+            .expect("parent node exists in the stack graph")
+    }
+
+    // Saves additional binding information from the loaded graph (eg. selector,
+    // definiens, import/export nodes and parents).
+    fn load_additional_info(&mut self, node_ref: GraphNodeRef) -> Result<(), BuildError> {
         let node = &self.graph[node_ref];
-        let definiens_node = match node.attributes.get(DEFINIENS_NODE_ATTR) {
-            Some(Value::Null) => return Ok(()),
-            Some(definiens_node) => &self.graph[definiens_node.as_syntax_node_ref()?],
-            None => return Ok(()),
+        let node_handle = self.node_handle_for_graph_node(node_ref);
+        let stack_graph_node = &self.stack_graph[node_handle];
+
+        let parents = if let Some(parents) = node.attributes.get(PARENTS_ATTR) {
+            parents
+                .as_list()?
+                .iter()
+                .flat_map(|value| {
+                    value
+                        .as_graph_node_ref()
+                        .map(|id| self.node_handle_for_graph_node(id))
+                })
+                .flat_map(|parent| {
+                    // ensure parents are either definitions or references
+                    let parent_node = &self.stack_graph[parent];
+                    if !parent_node.is_definition() && !parent_node.is_reference() {
+                        Err(BuildError::InvalidParent(node_ref))
+                    } else {
+                        Ok(parent)
+                    }
+                })
+                .collect()
+        } else {
+            Vec::new()
         };
-        // Save the definiens CST cursor so our caller can extract the mapping
-        // to the stack graph node later
-        self.definiens.insert(node_handle, definiens_node.clone());
+
+        if stack_graph_node.is_definition() {
```

Comment:
I prefer a more functional approach like `selector_value.map(|selector| ...)` when I need this construct i.e. returning `None` in the e.se branch of a `Option` destructuring.

However this is just a comment, not a critique. I note you sometimes use those monadic operators in other code.

Reply:
Yes, I tried using `map` here, but it doesn't work because we have `?` operators to handle and propagate attribute errors when doing coercion to the expected types..